src/rtcModule/webrtcAdapter.cpp

```diff
@@ -78,171 +88,382 @@ unsigned long generateId()
+    if (!info)
     {
-        const uint32_t kSize = 256;
-        char name[kSize] = {0};
-        char id[kSize] = {0};
-        if (info->GetDeviceName(i, name, kSize, id, kSize) != -1)
-        {
-            std::string sId = id;
-            if (sId.find(":1") == std::string::npos)
-            {
-                devices.push_back(cricket::Device(name, id));
-            }
-            else
-            {
-                devices.insert(devices.begin(), cricket::Device(name, id));
-            }
-            RTCM_LOG_INFO("Video input device %d: '%s'", i+1, name);
-        }
+        return videoDevices;
     }
 
-    if (devices.size() == 0)
+    uint32_t numDevices = info->NumberOfDevices();
+    for (uint32_t i = 0; i < numDevices; i++)
     {
-        RTCM_LOG_WARNING("Not suitable input video devices were found. Default device added.");
-        devices.push_back(cricket::Device("default", "0"));
+        char deviceName[256];
+        char uniqueName[256];
+        std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> deviceInfo(webrtc::VideoCaptureFactory::CreateDeviceInfo());
+        deviceInfo->GetDeviceName(i, deviceName, sizeof(deviceName), uniqueName, sizeof(uniqueName));
+        videoDevices.insert(std::pair<std::string, std::string>(deviceName, uniqueName));
     }
+
+    return videoDevices;
 }
 
-template<>
-void InputDeviceShared<webrtc::VideoTrackInterface, webrtc::VideoTrackSourceInterface>::createSource()
+void CaptureModuleLinux::openDevice(const std::string &videoDevice)
 {
-    assert(!mSource.get());
-
-#ifndef __ANDROID__
-    cricket::WebRtcVideoDeviceCapturerFactory factory;
-    std::unique_ptr<cricket::VideoCapturer> capturer(
-        factory.Create(cricket::Device(mOptions->device.name, 0)));
-    if (!capturer)
+    mCameraCapturer = webrtc::VideoCaptureFactory::Create(videoDevice.c_str());
+    if (!mCameraCapturer)
     {
-        RTCM_LOG_WARNING("Could not create video capturer for device '%s'", mOptions->device.name.c_str());
         return;
     }
 
-    mCapturer = capturer.get();
-    mSource = gWebrtcContext->CreateVideoSource(capturer.release(),
-        &(mOptions->constraints));
-#else
-    JNIEnv *env;
-    MEGAjvm->AttachCurrentThread(&env, NULL);
-    rtc::Thread *currentThread = rtc::ThreadManager::Instance()->CurrentThread();
-    mSource = new rtc::RefCountedObject<webrtc::AndroidVideoTrackSource>(currentThread, env, surfaceTextureHelper, false);
-    rtc::scoped_refptr<webrtc::VideoTrackSourceProxy> proxySource = webrtc::VideoTrackSourceProxy::Create(currentThread, currentThread, mSource);
-    if (startVideoCaptureWithParametersMID)
+    mCameraCapturer->RegisterCaptureDataCallback(this);
+
+    std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> deviceInfo(webrtc::VideoCaptureFactory::CreateDeviceInfo());
+    webrtc::VideoCaptureCapability capabilities;
+    deviceInfo->GetCapability(mCameraCapturer->CurrentDeviceName(), 0, capabilities);
```

Comment:
Why don't we simply pass here `mCapabilities`?
```suggestion
    deviceInfo->GetCapability(mCameraCapturer->CurrentDeviceName(), 0, mCapabilities);
```
That way we only have to set the `videoType` and we would honor all the capabilities provided by `GetCapability()`, rather than request to `StartCapture()` with our hardcoded capabilities.

Reply:
The `mCapabilities` is already fixed to the wanted video resolution and fps. Only the `interlaced` capability needs to be the same than the supported by the device. The video format is preset to `I420` because that's the output video format, so we can avoid the conversion of frames from another format.